SpatialGDK/Source/SpatialGDK/Private/Interop/Startup/DefaultServerWorkerStartupHandler.cpp

```diff
@@ -0,0 +1,459 @@
+#include "Interop/GlobalStateManager.h"
+#include "LoadBalancing/AbstractLBStrategy.h"
+#include "Schema/ServerWorker.h"
+#include "Utils/EntityFactory.h"
+#include "Utils/InterestFactory.h"
+#include "Utils/SpatialStatics.h"
+
+DEFINE_LOG_CATEGORY_STATIC(LogSpatialStartupHandler, Log, All);
+
+namespace SpatialGDK
+{
+FSpatialServerStartupHandler::FSpatialServerStartupHandler(USpatialNetDriver& InNetDriver, const FInitialSetup& InSetup)
+	: Setup(InSetup)
+	, NetDriver(&InNetDriver)
+{
+}
+
+bool FSpatialServerStartupHandler::TryFinishStartup()
+{
+	if (Stage == EStage::CreateWorkerEntity)
+	{
+		if (!bCalledCreateEntity)
+		{
+			bCalledCreateEntity = true;
+			WorkerEntityCreator.Emplace(*NetDriver, *NetDriver->Connection);
+		}
+		WorkerEntityCreator->ProcessOps(GetOps());
+		if (WorkerEntityCreator->IsFinished())
+		{
+			WorkerEntityId = WorkerEntityCreator->GetWorkerEntityId();
+			Stage = EStage::WaitForWorkerEntities;
+		}
+	}
+
+	if (Stage == EStage::WaitForWorkerEntities)
+	{
+		TMap<Worker_EntityId_Key, const ComponentData*> WorkerComponents;
+
+		for (const auto& EntityData : GetCoordinator().GetView())
+		{
+			const ComponentData* WorkerComponent =
+				EntityData.Value.Components.FindByPredicate(ComponentIdEquality{ SpatialConstants::SERVER_WORKER_COMPONENT_ID });
+			if (WorkerComponent != nullptr)
+			{
+				WorkerComponents.Emplace(EntityData.Key, WorkerComponent);
+			}
+		}
+		if (WorkerComponents.Num() >= Setup.ExpectedServerWorkersCount)
+		{
+			ensureMsgf(WorkerComponents.Num() == Setup.ExpectedServerWorkersCount,
+					   TEXT("There should never be more server workers connected than we expect. Expected %d, actually have %d"),
+					   Setup.ExpectedServerWorkersCount, WorkerComponents.Num());
+			WorkerComponents.GetKeys(WorkerEntityIds);
+			Stage = EStage::WaitForGSMEntity;
+		}
+	}
+
+	if (Stage == EStage::WaitForGSMEntity)
+	{
+		const EntityViewElement* GlobalStateManagerEntity = GetCoordinator().GetView().Find(GetGSM().GlobalStateManagerEntityId);
+		if (GlobalStateManagerEntity != nullptr)
+		{
+			Stage = EStage::DeriveDeploymentRecoveryState;
+		}
+	}
+
+	if (Stage == EStage::DeriveDeploymentRecoveryState)
+	{
+		const EntityViewElement& GlobalStateManagerEntity = GetCoordinator().GetView().FindChecked(GetGSM().GlobalStateManagerEntityId);
+		const ComponentData* StartupActorManagerData = GlobalStateManagerEntity.Components.FindByPredicate(
+			ComponentIdEquality{ SpatialConstants::STARTUP_ACTOR_MANAGER_COMPONENT_ID });
+		if (StartupActorManagerData != nullptr)
+		{
+			if (Schema_GetBoolCount(StartupActorManagerData->GetFields(), SpatialConstants::STARTUP_ACTOR_MANAGER_CAN_BEGIN_PLAY_ID) != 0)
+			{
+				// StartupActorManager's CAN_BEGIN_PLAY is raised after startup is finished, and is false in the initial snapshot.
+				// If it's true at this point, then this worker is either recovering or loading from a non-initial snapshot.
+				bIsRecoveringOrSnapshot =
```

Comment:
This logic is new to this PR right?
I believe it will fail if we're doing server travel (see `ResetGSM`).

Reply:
No, it's straightened and renamed from the stuff we had in the GSM regarding StartupActorManager: https://github.com/spatialos/UnrealGDK/blob/90dd4741dbc366ecd5e28f347a3c98a758c92213/SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp#L109
This intricate logic been written down as a comment, but I think it's clearer this way.

Although I didn't really consider ServerTravel; I think it's officially not supported right now, isn't it?